formats/json/commonMain/src/kotlinx/serialization/json/internal/JsonPath.kt

```diff
@@ -0,0 +1,81 @@
+package kotlinx.serialization.json.internal
+
+import kotlinx.serialization.*
+import kotlinx.serialization.descriptors.*
+
+internal class JsonPath {
+    /*
+     * Serial descriptor OR map key.
+     * Map key can also implement SD, so it also has '-2' as current index
+     */
+    private var currentObjectPath = arrayOfNulls<Any?>(8)
+    private var currentIndices = IntArray(8)
+    private var currentDepth = -1
+
+    fun pushObject(sd: SerialDescriptor) {
+        if (++currentDepth == currentObjectPath.size) {
+            resize()
+        }
+        currentObjectPath[currentDepth] = sd
+        currentIndices[currentDepth] = -1
+    }
+
+    fun updateCurrentMapKey(key: Any?) {
+        if (currentIndices[currentDepth] != -2 && ++currentDepth == currentObjectPath.size) {
+            resize()
+        }
+        currentObjectPath[currentDepth] = key
+        currentIndices[currentDepth] = -2
+    }
+
+    fun updateCurrentIndex(index: Int) {
+        currentIndices[currentDepth] = index
+    }
+
+    fun pop() {
+        // When we are ending map, we pop the last key and the outer field as well
+        if (currentIndices[currentDepth] == -2) {
+            currentIndices[currentDepth] = -1
+            currentDepth--
+        }
+        // Against top-level maps
+        if (currentDepth != -1) currentDepth--
+    }
+
+    @OptIn(ExperimentalSerializationApi::class)
+    fun currentPath(): String {
+        if (currentDepth == -1) return ""
+        return buildString {
+            append("Root type: ")
+            append(currentObjectPath[0])
```

Comment:
`SerialDescriptor.toString` also prints all properties, which may cause output to be bloated. I'd think we better stick with serial name only here.